radixdlt-engine/src/main/java/com/radixdlt/atom/TxLowLevelBuilder.java

```diff
@@ -271,14 +259,17 @@ public TxLowLevelBuilder read(SubstateId substateId) {
   }
 
   public TxLowLevelBuilder localDown(int index) {
-    var particle = localUpParticles.remove(index);
-    if (particle == null) {
-      throw new IllegalStateException("Local particle does not exist: " + index);
-    }
+    validateSubstate(localUpParticles.remove(index), index);
     instruction(REInstruction.REMicroOp.LDOWN, Shorts.toByteArray((short) index));
     return this;
   }
 
+  private void validateSubstate(LocalSubstate substate, int index) {
```

Comment:
The method is used in two places (not just  with `remove()`, but with `get()` too) and index is necessary exclusively to generate a useful exception message.

Reply:
How about making it instead a `validateLocalUpSubstateExists(index)`, which is called as a precondition before the get/remove on the next line? Whilst we're at it, could we rename `localUpParticles` as `localUpSubstates`?

Reply:
I think it affects existing behavior because there will be two instead of one accesses to the `localUpSubstates` map.
Also, I've made `validateSubstate` static and renamed to `validateSubstatePresentAtIndex` and think that this should address remaining objections.